File: backend/app/schemas/security.py
"""
Pydantic schemas for Security Events, Auditing, and Session Management.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class SecuritySeverity(str, Enum):
    """Severity classification for security events."""

    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class SecurityEventItem(BaseModel):
    """Sanitized security event representation."""

    id: uuid.UUID
    action: str
    severity: str = "info"
    user_id: uuid.UUID | None = None
    user_email: str | None = None
    ip_address: str | None = None
    user_agent: str | None = None
    status: str = "success"  # "success" | "failure"
    resource_type: str | None = None
    resource_id: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class SecurityEventPagination(BaseModel):
    """Pagination metadata for security event queries."""

    page: int
    page_size: int
    total_items: int
    total_pages: int


class SecurityEventListResponse(BaseModel):
    """Response envelope for paginated security events."""

    success: bool = True
    data: list[SecurityEventItem]
    pagination: SecurityEventPagination


class SecuritySummaryData(BaseModel):
    """Aggregated security metrics over recent time windows."""

    total_events: int
    failed_logins_24h: int
    rate_limit_exceeded_24h: int
    unauthorized_attempts_24h: int
    critical_events_24h: int
    token_reuse_detected_24h: int
    active_sessions_count: int


class SecuritySummaryResponse(BaseModel):
    """Response envelope for security summary metrics."""

    success: bool = True
    data: SecuritySummaryData


class ActiveSessionItem(BaseModel):
    """Metadata representation of an active user session / refresh token."""

    id: uuid.UUID
    user_id: uuid.UUID
    created_at: datetime
    expires_at: datetime
    is_current: bool = False


class ActiveSessionListResponse(BaseModel):
    """Response envelope for active user sessions."""

    success: bool = True
    data: list[ActiveSessionItem]

File: frontend/types/security.ts
/**
 * TypeScript types for Security Events, Health, and System Observability.
 */

export type SecuritySeverity = "info" | "warning" | "critical";

export interface SecurityEvent {
  id: string;
  action: string;
  severity: SecuritySeverity;
  user_id: string | null;
  user_email: string | null;
  ip_address: string | null;
  user_agent: string | null;
  status: "success" | "failure";
  resource_type: string | null;
  resource_id: string | null;
  details: Record<string, unknown>;
  created_at: string;
}

export interface SecurityEventPagination {
  page: number;
  page_size: number;
  total_items: number;
  total_pages: number;
}

export interface SecurityEventListResponse {
  success: boolean;
  data: SecurityEvent[];
  pagination: SecurityEventPagination;
}

export interface SecuritySummaryData {
  total_events: number;
  failed_logins_24h: number;
  rate_limit_exceeded_24h: number;
  unauthorized_attempts_24h: number;
  critical_events_24h: number;
  token_reuse_detected_24h: number;
  active_sessions_count: number;
}

export interface SecuritySummaryResponse {
  success: boolean;
  data: SecuritySummaryData;
}

export interface ActiveSession {
  id: string;
  user_id: string;
  created_at: string;
  expires_at: string;
  is_current?: boolean;
}

export interface DependencyHealth {
  status: "healthy" | "degraded" | "unhealthy";
  latency_ms?: number;
  details?: Record<string, unknown>;
}

export interface SystemMetrics {
  status: string;
  uptime_seconds: number;
  memory_usage_mb: number;
  environment: string;
  database_status: string;
  database_latency_ms: number | null;
  redis_status: string;
  redis_latency_ms: number | null;
  celery_status: string;
  ai_status: string;
  faiss_vectors: number;
  total_users_count: number;
  total_documents_count: number;
  total_workflows_count: number;
  total_executions_count: number;
  timestamp: string;
}

export interface SystemMetricsResponse {
  success: boolean;
  data: SystemMetrics;
}
